dll/kernel32.cpp

```diff
@@ -198,32 +202,106 @@ namespace kernel32 {
 		exit(uExitCode);
 	}
 
-	int WIN_FUNC CreateProcessA(
-		const char *lpApplicationName,
-		char *lpCommandLine,
-		void *lpProcessAttributes,
-		void *lpThreadAttributes,
-		int bInheritHandles,
-		int dwCreationFlags,
-		void *lpEnvironment,
-		const char *lpCurrentDirectory,
-		void *lpStartupInfo,
-		void *lpProcessInformation
-	) {
-		printf("CreateProcessA %s \"%s\" %p %p %d 0x%x %p %s %p %p\n",
-			lpApplicationName,
-			lpCommandLine,
-			lpProcessAttributes,
+	BOOL WIN_FUNC GetExitCodeProcess(HANDLE hProcess, LPDWORD lpExitCode) {
+		DEBUG_LOG("GetExitCodeProcess\n");
+
+		processes::Process* process = processes::processFromHandle(hProcess, false);
+		*lpExitCode = process->exitCode;
+		return 1; // success in retrieval
+	}
+
+	struct PROCESS_INFORMATION {
+		HANDLE hProcess;
+		HANDLE hThread;
+		DWORD  dwProcessId;
+		DWORD  dwThreadId;
+	};
+
+
+ 	BOOL WIN_FUNC CreateProcessA(
+		LPCSTR lpApplicationName,
+		LPSTR lpCommandLine,
+ 		void *lpProcessAttributes,
+ 		void *lpThreadAttributes,
+		BOOL bInheritHandles,
+		DWORD dwCreationFlags,
+		LPVOID lpEnvironment,
+		LPCSTR lpCurrentDirectory,
+ 		void *lpStartupInfo,
+		PROCESS_INFORMATION *lpProcessInformation
+ 	) {
+		DEBUG_LOG("CreateProcessA %s \"%s\" %p %p %d 0x%x %p %s %p %p\n",
+ 			lpApplicationName,
+ 			lpCommandLine,
+ 			lpProcessAttributes,
 			lpThreadAttributes,
 			bInheritHandles,
 			dwCreationFlags,
 			lpEnvironment,
-			lpCurrentDirectory ? lpCurrentDirectory : "<none>",
-			lpStartupInfo,
-			lpProcessInformation
-		);
-		printf("Cannot handle process creation, aborting\n");
-		exit(1);
+			lpCurrentDirectory,
+ 			lpStartupInfo,
+ 			lpProcessInformation
+ 		);
+
+		// Argument parsing
+		// First: how many arguments do we have?
+		size_t argc = 2;
+
+		for (size_t i = 1; i < strlen(lpCommandLine); i++) {
+			if (isspace(lpCommandLine[i]) && !isspace(lpCommandLine[i - 1]))
+				argc++;
+		}
+
+		char **argv = (char **) calloc(argc + 1, sizeof(char*));
+		argv[0] = wibo::executableName;
+		argv[1] = (char *) files::pathFromWindows(lpApplicationName).string().c_str();
+
+		char* arg = strtok(lpCommandLine, " ");
+		size_t current_arg_index = 2;
+
+		while (arg != NULL) {
+			// We're deliberately discarding the first token here
+			// to prevent from doubling up on the target executable name
+			// (it appears as lpApplicationName, and as the first token in lpCommandLine)
+			arg = strtok(NULL, " ");
+			argv[current_arg_index++] = arg;
+		}
+
+		argv[argc] = NULL; // Last element in argv should be a null pointer
+
+		// YET TODO: take into account process / thread attributes, environment variables
+		// working directory, etc.
+		setenv("WIBO_DEBUG_INDENT", std::to_string(wibo::debugIndent + 1).c_str(), true);
```

Comment:
By this do you mean environment variables for the process to be spawned?

Reply:
yup! parsing and mashing lpEnvironment into a format posix_spawn likes.